### ISSUE TITLE
损失函数

### ISSUE THREAD

**Issue by alasq** [alasq/tensorflow-learning#3]
# 经典损失函数
在解决多分类问题,我们会设置n个输出节点,理想状态下,我们识别一个样本数据类别k,那么这个类型的节点输出是1,其他节点输出是0,整体输出应该类似`[0,1,0,0,0,0,0,0]` ,而实际输出应该都是0到1之间的数。
原始结果肯定有不在0到1之间的。我们通过 softmax回归层把输出变换到0到1之间

## 分类问题一般采用交叉熵作为损失函数
交叉熵刻画的是两个概率分布之间的距离
`H((1,0,0),(0.5,0.4,0.1))=-(0*log0.5+1*log0.4+0*log0.1)≈0.3 `
```python
cross_entropy= -tf.reduce_mean(y_ * tf.log(tf.clip_by_value(y,1e-10,1.0)))
```

tf.clip_by_value(y,1e-10,1.0) 函数y里面的值,转换成1e-10 到1.0 之间的值
tf.log(x) 将x取对数
tf.reduce_mean 相加

## 回归问题一般采用 均方误差 MES (mean squared error)作为损失函数
<a href="https://www.codecogs.com/eqnedit.php?latex=MSE(y,y')=\frac{\sum_{i=1}^{n}(y_{i}-{y_{i}'})^{2}}{n}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?MSE(y,y')=\frac{\sum_{i=1}^{n}(y_{i}-{y_{i}'})^{2}}{n}" title="MSE(y,y')=\frac{\sum_{i=1}^{n}(y_{i}-{y_{i}'})^{2}}{n}" /></a>

## 自定义损失函数
假如预测商品销量，预测值比实际值多的话，会损失商品的成本，预测值比实际值少的话，就是损失销量。假设成本是10元，销售利润是1元。
<a href="https://www.codecogs.com/eqnedit.php?latex=Loss(y,y')=\sum_{i=1}^{n}f(y_{i},y_{i}')&space;,&space;f(x,y)=\left\{\begin{matrix}&space;10(x-y)&space;&&space;x>y\\&space;1(y-x)&space;&&space;x\leq&space;y&space;\end{matrix}\right." target="_blank"><img src="https://latex.codecogs.com/gif.latex?Loss(y,y')=\sum_{i=1}^{n}f(y_{i},y_{i}')&space;,&space;f(x,y)=\left\{\begin{matrix}&space;10(x-y)&space;&&space;x>y\\&space;1(y-x)&space;&&space;x\leq&space;y&space;\end{matrix}\right." title="Loss(y,y')=\sum_{i=1}^{n}f(y_{i},y_{i}') , f(x,y)=\left\{\begin{matrix} 10(x-y) & x>y\\ 1(y-x) & x\leq y \end{matrix}\right." /></a>

用tf来标识这个自定义的损失函数
```py
loss=tf.reduce_sum(tf.where(
tf.greater(v1,v2),
(v1-v2)*a,
(v2-v1)*b
))
```
